Improve documentation api examples

engelmi opened the pull request:
This PR improves the documentation with the API examples as well as about the code generation based on the introspection data. 
It contains the following changes: 
- introducing a new theme to support the tabs for showing the examples in different programming languages
- adding a custom js script was added to synchronize the tabs (e.g. when `Python` tab is clicked for the setup, all other tabs follow suit). 
- adding Go examples and aligning the output with the python ones
- keeping the Rust and C/C++ examples, but not showing them in the documentation for now as only a subset of the 9 use cases are written in those languages - those can be easily added later on (future work)

The changes can be viewed by running mkdocs locally described in https://github.com/containers/bluechi/blob/main/doc/README.md

Some images to showcase the theme change:

![image](https://github.com/containers/bluechi/assets/26504678/3a2a73a2-9676-4758-84fb-5a0117dd31ea)

![image](https://github.com/containers/bluechi/assets/26504678/77561881-4367-4b1c-ab1a-f954ee94374d)

engelmi committed:
add tab feature to mkdocs

By using tabs, the code snippet for the api examples can
be grouped by use case and split by language.
Unfortunately, the pymdownx.tabbed seems to not work
properly with the readthedocs theme. Because of that the
theme was exchanged with material. The dependency for it
was added.
In order to sync tabs with the same title, e.g. multiple
tabs for python code, a dedicated js script was added.

Signed-off-by: Michael Engel <mengel@redhat.com>

mwperina commented on doc/api-examples/c/setup.md:
```diff
@@ -0,0 +1,37 @@
+apt install build-essential 
+# on Fedora/CentOS
+dnf install systemd-devel
+dnf install gcc g++
```
Wouldn't it be easier to install all requirements in a signle command?

```
apt install libsystemd-dev build-essential
```
```
dnf install gcc g++ systemd-devel
```

engelmi replied:
Good point. I'll change it.

mwperina commented on doc/api-examples/c/setup.md:
```diff
@@ -0,0 +1,37 @@
+
+**Build and run the sample project**
+
+The snippet chosen to run needs to be copied into a file, e.g. with name `bluechi.c`, in the `bluechi` directory. After that the example can be build and run via:
```
I'd rather use `bluechi-sample.c` to prevent confusion with `bluechi` daemon. But to align with python examples, wouldn't it be better to have each example executable (with `main()` function included)? Then instead of `bluechi.c` we could use `SAMPLE.c` or `<SAMPLE.c>`

engelmi replied:
> have each example executable (with main() function included)? 

That is actually the case. 
I agree regarding the setup description here. I changed it to refer to the `<filename>` instead. Could you have another look? @mwperina

mwperina commented on doc/api-examples/go/setup.md:
Should we also here suggest install required Golang related packages?

engelmi replied:
Which packages do you mean? The only library that needs to be installed is the godbus/dbus - and this is described a bit below.

mwperina replied:
Not sure about additional packages, but at least following should be installed, right?
```
dnf install golang
```

mwperina replied:
And maybe also `golang-github-godbus-dbus-devel` package

engelmi replied:
Yes, you are right - since we describe how to install `gcc`/`g++` it makes sense to briefly show how to install go - and rust and python probably. However, I'd only link to the the documentation on how to set them up, e.g. https://go.dev/doc/install

Regarding the godbus package: Go, Rust and Python provide package managers out of the box which are widely used, so I'd rather use those. Also, a `go get` or `cargo add` adds the dependency to the project file - not sure if `dnf` does this.

mwperina replied:
OK, thanks

engelmi committed:
restructure the api and bluechi examples documentation

Group the api examples by use case and provide them for each
language we want to show an example for. Also, instead of
only showing snippets, describe a basic setup to run the
example code. Adding all examples for Go and a few for C and
Rust as well.

Signed-off-by: Michael Engel <mengel@redhat.com>

mwperina commented on doc/api-examples/go/setup.md:
````diff
@@ -0,0 +1,30 @@
+
+```bash
+go get github.com/godbus/dbus/v5
+```
````
Wouldn't it be better to just install existing RPM `golang-github-godbus-dbus-devel`?

engelmi replied:
responded https://github.com/containers/bluechi/pull/563#discussion_r1325752617

mwperina commented on doc/api-examples/python/setup.md:
````diff
@@ -0,0 +1,16 @@
+The python snippets require the [dasbus](https://dasbus.readthedocs.io/en/latest/) package. Install it via:
+
+```bash
+pip install dasbus
````
Can't we use RPM package instead?

engelmi replied:
responded https://github.com/containers/bluechi/pull/563#discussion_r1325752617

mwperina commented on doc/api-examples/rust/setup.md:
````diff
@@ -0,0 +1,36 @@
+The Rust snippets require the [dbus](https://docs.rs/dbus/latest/dbus/index.html) create. Navigate into the `bluechi` directory and add it as a dependency via:
+
+```bash
+cargo add dbus
````
Wouldn't it be better to use

```
dnf install rust-dbus-devel
```

engelmi replied:
responded https://github.com/containers/bluechi/pull/563#discussion_r1325752617

mwperina reviewed (approved):
LGTM
